File: javascript/234.palindrome-linked-list.js
/*
 * @lc app=leetcode id=234 lang=javascript
 *
 * [234] Palindrome Linked List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {boolean}
 */
var isPalindrome = function (head) {
  let slow = head,
    fast = head;

  // Solution 1: O(1) space
  while (fast && fast.next) {
    slow = slow.next;
    fast = fast.next.next;
  }
  let tail = null;
  while (slow) {
    const temp = slow;
    slow = slow.next;
    temp.next = tail;
    tail = temp;
  }
  while (tail) {
    if (head.val !== tail.val) return false;
    head = head.next;
    tail = tail.next;
  }

  // Solution 2: array O(n) space
  // const nums = [];
  // while (fast && fast.next) {
  //   nums.push(slow.val);
  //   slow = slow.next;
  //   fast = fast.next.next;
  // }
  // if (fast !== null) slow = slow.next;
  // while (slow) {
  //   if (slow.val !== nums.pop()) return false;
  //   slow = slow.next;
  // }

  return true;
};
// @lc code=end

File: rust/234.palindrome-linked-list.rs
/*
 * @lc app=leetcode id=234 lang=rust
 *
 * [234] Palindrome Linked List
 */

// @lc code=start
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
//
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn is_palindrome(head: Option<Box<ListNode>>) -> bool {
        let mut nums = Vec::new();
        let (mut slow, mut fast) = (&head, &head);
        loop {
            match (slow.as_ref(), fast.as_ref()) {
                (Some(s_node), Some(f_node)) => {
                    if let Some(f_next_node) = f_node.next.as_ref() {
                        nums.push(s_node.val);
                        slow = &s_node.next;
                        fast = &f_next_node.next;
                    } else {
                        break;
                    }
                }
                _ => break,
            }
        }
        if fast.is_some() {
            slow = &slow.as_ref().unwrap().next;
        }
        while nums.len() > 0 {
            match (nums.pop(), slow.as_ref()) {
                (Some(num), Some(s_node)) => {
                    if num != s_node.val {
                        return false;
                    }
                    slow = &s_node.next;
                }
                _ => break,
            }
        }
        return true;
    }
}
// @lc code=end
